src/db_bench/db_bench.rs

```diff
@@ -0,0 +1,303 @@
+use slatedb::db::Db;
+use std::collections::VecDeque;
+use std::sync::{Arc, Mutex};
+use std::time::Duration;
+use tokio::time::Instant;
+
+pub trait KeyGenerator: Send {
+    fn next_key(&mut self) -> Bytes;
+}
+
+// TODO: implement other distributions
+
+pub struct RandomKeyGenerator {
+    key_bytes: usize,
+    rng: XorShiftRng,
+}
+
+impl RandomKeyGenerator {
+    pub fn new(key_bytes: usize) -> Self {
+        Self {
+            key_bytes,
+            rng: rand_xorshift::XorShiftRng::from_entropy(),
+        }
+    }
+}
+
+impl KeyGenerator for RandomKeyGenerator {
+    fn next_key(&mut self) -> Bytes {
+        let mut bytes = vec![0u8; self.key_bytes];
+        self.rng.fill_bytes(bytes.as_mut_slice());
+        Bytes::copy_from_slice(bytes.as_slice())
+    }
+}
+
+pub struct DbBench {
+    key_gen_supplier: Box<dyn Fn() -> Box<dyn KeyGenerator>>,
+    val_size: usize,
+    write_options: WriteOptions,
+    write_rate: Option<u32>,
+    #[allow(dead_code)]
+    read_rate: Option<u32>,
+    write_tasks: u32,
+    #[allow(dead_code)]
+    read_tasks: u32,
+    num_rows: Option<u64>,
+    duration: Option<Duration>,
+    db: Arc<Db>,
+}
+
+impl DbBench {
+    #[allow(clippy::too_many_arguments)]
+    pub fn write(
+        key_gen_supplier: Box<dyn Fn() -> Box<dyn KeyGenerator>>,
+        val_size: usize,
+        write_options: WriteOptions,
+        write_rate: Option<u32>,
+        write_tasks: u32,
+        num_rows: Option<u64>,
+        duration: Option<Duration>,
+        db: Arc<Db>,
+    ) -> Self {
+        Self {
+            key_gen_supplier,
+            val_size,
+            write_options,
+            write_rate,
+            read_rate: None,
+            write_tasks,
+            read_tasks: 0,
+            num_rows,
+            duration,
+            db,
+        }
+    }
+
+    pub async fn run(&self) {
+        let rate_limiter = self.write_rate.map(|r| {
+            Arc::new(
+                RateLimiter::builder()
+                    .initial(r as usize)
+                    .max(r as usize)
+                    .interval(Duration::from_millis(1))
+                    .refill((r / 1000) as usize)
+                    .build(),
+            )
+        });
+        let stats_recorder = Arc::new(StatsRecorder::new());
+        let mut write_tasks = Vec::new();
+        for _ in 0..self.write_tasks {
+            let mut write_task = WriteTask::new(
+                (*self.key_gen_supplier)(),
+                self.val_size,
+                self.write_options.clone(),
+                self.num_rows,
+                self.duration,
+                rate_limiter.clone(),
+                stats_recorder.clone(),
+                self.db.clone(),
+            );
+            write_tasks.push(tokio::spawn(async move { write_task.run().await }));
+        }
+        tokio::spawn(async move { dump_stats(stats_recorder).await });
+        for write_task in write_tasks {
+            write_task.await.unwrap();
```

Comment:
Nit: I think you can use `futures::future::join_all` here if you want

Reply:
eh I still have to go over the results and unwrap them all